### PR TITLE
BUG: align with broadcast_axis, #13194

### LINKED ISSUE #13194: align with broadcast_axis specified always uses inner join when aligning dataframe and series on other axis.
Pandas version 0.18.1:

``` python
In[1]: import pandas as pd
In[2]: df = pd.DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
In[3]: ts = pd.Series([5., 6., 7.])
In[4]: df.align(ts, join='outer', axis=0, broadcast_axis=1)
Out[4]: 
(     A    B
 0  1.0  2.0
 1  3.0  4.0,      A    B
 0  5.0  5.0
 1  6.0  6.0)
```

However, as 'join' is specified to be 'outer', I would have expected the following result:

``` python
Out[4]: 
(     A    B
 0  1.0  2.0
 1  3.0  4.0
 2  NaN  NaN,      A    B
 0  5.0  5.0
 1  6.0  6.0
 2  7.0  7.0
 dtype: float64)
```

Same problem occurs for:

``` python
df.align(ts, join='right', axis=0, broadcast_axis=1)
ts.align(df, join='outer', axis=0, broadcast_axis=1)
ts.align(df, join='left', axis=0, broadcast_axis=1)
```

### PR DESCRIPTION
**Pull request by pfrcks**
Closes #13194 
- Implemented the _easy fix_
- Future work : To implement broadcast axis remaining functionality.
- All tests in test_generic.py passed.

### CODE REVIEW

**Comment by jreback** [comment]
tests

we r on 0.18.2

**Comment by pfrcks** [comment]
Hi,

I seem to have made the documentation in the wrong file. I will correct it and push the changes.
Rest all is OK, right?
This is my first pull request to pandas, sorry for the noob errors.

**Comment by pfrcks** [comment]
@jreback I have corrected the documentation to the proper version file. Is this everything?
I have ran the tests also.
Am I missing something?

**Comment by jreback** [comment]
you need to add tests which fail before this patch and pass after

**Comment by kdebrab** [comment]
Apart from the indispensable unit test, I'd propose to move [the lines](https://github.com/pydata/pandas/blob/master/pandas/core/generic.py#L4172):
```python
if axis is not None:
    axis = self._get_axis_number(axis)
```
to right before the `if broadcast_axis == 1` block, such that broadcasting will work also in case align is called with `axis='index'` instead of `axis=0`.

Thanks!

**Comment by pfrcks** [comment]
@jreback @kdebrab I've included the unit test and have also moved the location of the axis check.
Can you please take a look?

File: pandas/tests/test_generic.py [hunk 1]
```diff
@@ -1874,6 +1874,52 @@ def test_pipe_panel(self):
         with tm.assertRaises(ValueError):
             result = wp.pipe((f, 'y'), x=1, y=1)
 
+    def test_align_joins(self):
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
```
Comment (jreback) [inline pandas/tests/test_generic.py]: add the issue number as a comment here

rename to ``test_align_broadcast_axis``

File: pandas/tests/test_generic.py [hunk 2]
```diff
@@ -1874,6 +1874,52 @@ def test_pipe_panel(self):
...
+        ts = Series([5., 6., 7.])
+        check = df.align(ts, join='outer', axis=0, broadcast_axis=1)
+        df1 = DataFrame(check[0])
```
Comment (jreback) [inline pandas/tests/test_generic.py]: all of these tests need to go in ``tests/frame/test_axis_select_reinde`` and/or ``tests/series/test_indexing``. right after the other align tests.

Comment (jreback) [inline pandas/tests/test_generic.py]: model them after the existing tests.

File: doc/source/whatsnew/v0.18.2.txt [hunk 3]
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. (:issue:`13194`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: be more descriptive of what you fixed here. remember a user is reading this.

**Comment by pfrcks** [comment]
@jreback Have made the required changes. 
Please comment if something's lacking.

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 4]
```diff
@@ -880,3 +880,53 @@ def test_reindex_multi(self):
...
+                                        [pd.np.nan, pd.np.nan]]),
+                              columns=list('AB'))
+        expected2 = DataFrame(np.array([[5., 5.], [6., 6.], [7., 7.]]),
+                              columns=list('AB'))
+        assert_frame_equal(result1, expected1)
+        assert_frame_equal(result2, expected2)
+
+        # For 'inner' join
+        result = df.align(ts, join='inner', axis=0, broadcast_axis=1)
+        result1 = DataFrame(result[0])
```
Comment (sinhrks) [inline pandas/tests/frame/test_axis_select_reindex.py]: Not needed. Test returned obj (`result[0]` and `[1]` as it is.

**Comment by sinhrks** [comment]
Can u add tests with:
- alternately different `Index` to clheck left and right join.
- `Series.align(DataFrame)`

File: pandas/tests/frame/test_axis_select_reindex.py (hunk 4 above)
Comment (pfrcks) [inline pandas/tests/frame/test_axis_select_reindex.py]: @sinhrks I'm not sure that I understand. Can you please elaborate?

**Comment by pfrcks** [comment]
@sinhrks Sure. It's late night here right now. I will do this in the morning.

File: pandas/tests/frame/test_axis_select_reindex.py (hunk 4 above)
Comment (sinhrks) [inline pandas/tests/frame/test_axis_select_reindex.py]: Pls remove `DataFrame(result..)` like below.

```
        result = df.align(...)
        expected1 = ...
        expected2 = ...
        assert_frame_equal(result[0], expected1)
        assert_frame_equal(result[1], expected2)
```

**Comment by pfrcks** [comment]
@sinhrks I've made some changes. Please take a look and comment. This is what you meant right?

File: doc/source/whatsnew/v0.18.2.txt [hunk 5]
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. Earlier, upon supplying value of join argument as any of the four('outer', 'inner', 'left', 'right' ), align() was giving erraneous output. Align with broadcast_axis specified was using 'inner' join consistently irrespective of the value of 'join' provided when aligning dataframe and series on other axis. The problem was identified to be in pandas/core/generic.py and has been subsequently fixed. (:issue:`13194`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: Single line of comments pls. these are read by the users and don't need this kind of explanation.

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 6]
```diff
@@ -880,3 +880,66 @@ def test_reindex_multi(self):
...
+        # For 'left' join on different axis
+        result = df.align(ts, join='left', axis=1, broadcast_axis=1)
+        expected1 = DataFrame(np.array([[1., 2.], [3., 4.]]),
+                              columns=list('AB'))
+        expected2 = DataFrame(np.array([[5., 5.], [6., 6.], [7., 7.]]),
+                              columns=list('AB'))
+        assert_frame_equal(result[0], expected1)
+        assert_frame_equal(result[1], expected2)
+
+        # Series.align(DataFrame) tests, 'outer' join
```
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: these 2 series-frame joins should be in ``tests/series/test_indexing.py``

**Comment by pfrcks** [comment]
@jreback Done. Kindly check and comment.

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 7]
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
...
         assert_frame_equal(result, expected)
+
+    def test_align_broadcast_axis(self):
+        # GH 13194
+        # First four tests for DataFrame.align(Index)
+        # For 'right' join
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
+        ts = Series([5., 6., 7.])
+
+        result = df.align(ts, join='right', axis=0, broadcast_axis=1)
```
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: try one of these with ``broadcast_axis='columns'`` and one with ``broadcast_axis='index'`` as well

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 8]
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
...
+        expected1 = DataFrame(np.array([[1., 2.], [3., 4.],
```
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: call these ``left`` and ``right`` (and same below)

File: pandas/tests/frame/test_axis_select_reindex.py (hunk 7 above)
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: ``left, right = df.align(....)`` and same below

File: pandas/tests/series/test_indexing.py [hunk 9]
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
         result2 = s.ix['foo']
         self.assertEqual(result.name, s.name)
         self.assertEqual(result2.name, s.name)
+
+    def test_align_broadcast_axis_series(self):
+        # GH 13194
+        # Series.align(DataFrame) tests, 'outer' join
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
+        ts = Series([5., 6., 7.])
+        result = ts.align(df, join='outer', axis=0, broadcast_axis=1)
```
Comment (jreback) [inline pandas/tests/series/test_indexing.py]: same name scheme here

File: pandas/tests/series/test_indexing.py [hunk 10]
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
...
+        expected1 = DataFrame(np.array([[5., 5.], [6., 6.], [7., 7.]]),
+                              columns=list('AB'))
+        expected2 = DataFrame(np.array([[1., 2.], [3., 4.],
+                                        [pd.np.nan, pd.np.nan]]),
+                              columns=list('AB'))
+        self.assertTrue(result[0].equals(expected1))
```
Comment (jreback) [inline pandas/tests/series/test_indexing.py]: ``tm.assert_frame_equal``

File: doc/source/whatsnew/v0.18.2.txt [hunk 11]
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with broadcast_axis. (:issue:`13194`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: ``Bug in ``

use double-backticks around ``broadcast_axis``.

say ``incorrect output`` (not wrong output)

Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: say that this was not taking into account the ``how`` parameter when ``broadcast_axis`` was passed

File: pandas/tests/frame/test_axis_select_reindex.py [hunk 12]
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
...
+                                        [pd.np.nan, pd.np.nan]]),
+                              columns=list('AB'))
+        expected2 = DataFrame(np.array([[5., 5.], [6., 6.], [7., 7.]]),
+                              columns=list('AB'))
+        assert_frame_equal(result[0], expected1)
+        assert_frame_equal(result[1], expected2)
+
+        # For 'right' join on different index
+        result = df.align(ts, join='right', axis=1, broadcast_axis=1)
+        expected1 = DataFrame(np.array([[1., 2.], [3., 4.]]),
```
Comment (jreback) [inline pandas/tests/frame/test_axis_select_reindex.py]: you don't need to pass these as numpy arrays, just list-of-lists is fine

**Comment by pfrcks** [comment]
@jreback 

        right = df.align(ts, join='right', axis=1, broadcast_axis='index')
        expected1_right = DataFrame([[pd.np.nan, pd.np.nan, pd.np.nan],
                                     [pd.np.nan, pd.np.nan, pd.np,nan]],
                                    columns=[0, 1, 2])
        expected2_right = Series([5., 6., 7.])
        assert_frame_equal(right[0], expected1_right)
        assert_series_equal(right[1], expected2_right)  

While trying to run the above code, nosetests gives the error 

AssertionError: Series.index are different

        Series.index classes are not equivalent
        [left]:  Index([u'A', u'B'], dtype='object')
        [right]: RangeIndex(start=0, stop=2, step=1)

However, upon running in the interpreter

>>> right = df.align(ts, join='right', axis=1, broadcast_axis='index')
>>> right[1]
0    5.0
1    6.0
2    7.0
dtype: float64
>>> right[1].index
RangeIndex(start=0, stop=3, step=1)
>>> expected2_right = Series([5., 6., 7.])
>>> expected2_right.index
RangeIndex(start=0, stop=3, step=1)


We find that both indexes are same.
I am currently in pandas_dev environment.
Is this something assert_series_frame is doing?

File: pandas/tests/series/test_indexing.py [hunk 13]
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
...
+        self.assertTrue(result[1].equals(expected2))
+
+        # Series.align(DataFrame) tests, 'inner' join
+        result = ts.align(df, join='inner', axis=0, broadcast_axis=1)
```
Comment (sinhrks) [inline pandas/tests/series/test_indexing.py]: we can make all tests more comprehensive as "inner" and "outer" outputs the same, and "left" and "right" outputs inversed results.

```
result = df.align(series, ...)
self.assert_frame_equal(result[0], expected1)
self.assert_frame_equal(result[1], expected2)
result = series.align(df, ...)
self.assert_frame_equal(result[0], expected2)
self.assert_frame_equal(result[1], expected1)
```

Comment (sinhrks) [inline pandas/tests/series/test_indexing.py]: missed @jreback suggestion for variable names:

```
left, right = df.align(series, ...)
self.assert_frame_equal(left, expected_left)
self.assert_frame_equal(right, expected_right)
left, right = series.align(df, ...)
self.assert_frame_equal(right, expected_left)
self.assert_frame_equal(left, expected_right)
```

Comment (pfrcks) [inline pandas/tests/series/test_indexing.py]: @jreback Can you please look into the issue I have reported above.

Comment (jreback) [inline pandas/tests/series/test_indexing.py]: a single Ping is enough then you need to wait for comments

**Comment by jreback** [comment]
can you rebase / update?

**Comment by pfrcks** [comment]
@jreback I'm not sure I understand. I have rebased my_branch with master.

**Comment by jreback** [comment]
doesn't look like it
you need to push

**Comment by pfrcks** [comment]
@jreback 
I have used the following commands:

`git checkout my_branch`
`git rebase master`
`git push -u origin my_branch`

**Comment by pfrcks** [comment]
Sorry I closed my mistake

**Comment by jreback** [comment]
got rebase origin/master

**Comment by jreback** [comment]
got -> git

**Comment by jreback** [comment]
you are rebasing to a local master which is not updated

**Comment by pfrcks** [comment]
@jreback 

I followed the steps:

1) `git remote add ups https://github.com/pydata/pandas.git`
2) `git fetch ups`
3) `git checkout master`
4) `git rebase ups/master`

> First, rewinding head to replay your work on top of it...
Applying: BUG: GH13194 fixed
Applying: DOC : Change in release notes
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.18.2.txt
<stdin>:19: trailing whitespace.
- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. (:issue:`13194`) 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): doc/source/whatsnew/v0.18.2.txt deleted in HEAD and modified in DOC : Change in release notes. Version DOC : Change in release notes of doc/source/whatsnew/v0.18.2.txt left in tree.
Failed to merge in the changes.
Patch failed at 0002 DOC : Change in release notes
The copy of the patch that failed is found in:
   /home/pfrcks/issues/pandas-pfrcks/.git/rebase-apply/patch
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

I removed the whitespace error and then synced the changes but still I am getting the same issue.

**Comment by jreback** [comment]
closing, but if you want to update / rebase. pls reopen

**Pull request closed without merging**